ldms/src/sampler/zfs_leafvdevs/zfs_leafvdevs.c

```diff
@@ -0,0 +1,496 @@
+
+	sampler_base = base_config(avl, SAMP, "zfs_leafvdevs", log_fn);
+	if ((g_zfs = libzfs_init()) == NULL) {
+		rc = errno;
+		ldmsd_log(LDMSD_LERROR,
+			  SAMP " : Failed to initialize libzfs: %d\n", errno);
+		ldmsd_log(LDMSD_LERROR,
+			  SAMP
+			  " : Is the zfs module loaded or zrepl running?\n");
+	} else {
+		rc = initialize_ldms_structs();
+	}
+
+	if (rc < 0) {
+		base_del(sampler_base);
+		sampler_base = NULL;
+	}
+
+	return rc;
+}
+
+/*****************************************************************************
+ * WHAT:
+ * reallocate heap size plus 1 zpool struct and one vdev struct
+ * CALLER:
+ * self, (plugin)
+ ****************************************************************************/
+static int resize_metric_set()
+{
+	size_t previous_heap_size;
+	size_t new_heap_size;
+	int rc = 0;
+
+	previous_heap_size = ldms_set_heap_size_get(sampler_base->set);
+	base_set_delete(sampler_base);
+
+	new_heap_size = previous_heap_size;
+	new_heap_size += zpool_vdev_heap_sz;
+
+	if (base_set_new_heap(sampler_base, new_heap_size) == NULL) {
+		rc = errno;
+		ldmsd_log(LDMSD_LERROR,
+			  SAMP " : Failed to resize metric set heap: %d\n",
+			  errno);
+	} else {
+		log_fn(LDMSD_LDEBUG, "ldms resize of list successful\n");
+	}
+	return rc;
+}
+
+static int sample(struct ldmsd_sampler *self)
+{
+	int rc = 0;
+
+	base_sample_begin(sampler_base);
+
+	list_handle =
+	    ldms_metric_get(sampler_base->set, index_store.vdev_list_idx);
+	ldms_list_purge(sampler_base->set, list_handle);
+
+	rc = zpool_iter(g_zfs, get_leaf_stats, NULL);
+	if (rc != 0) {
+		log_fn(LDMSD_LERROR,
+		       SAMP " sample():zfs_pool print_stat() failed: %d\n", rc);
+	}
+
+	base_sample_end(sampler_base);
+
+	return rc;
+}
+
+static void term(struct ldmsd_plugin *self)
+{
+	log_fn(LDMSD_LDEBUG, SAMP " term() called\n");
+	base_set_delete(sampler_base);
+	base_del(sampler_base);
+	sampler_base = NULL;
+}
+
+static ldms_set_t get_set(struct ldmsd_sampler *self)
+{
+	return NULL;
+}
+
+static const char *usage(struct ldmsd_plugin *self)
+{
+	log_fn(LDMSD_LDEBUG, SAMP " usage() called\n");
+	return "config name=" SAMP " " BASE_CONFIG_SYNOPSIS BASE_CONFIG_DESC;
+}
+
+struct ldmsd_plugin *get_plugin(ldmsd_msg_log_f pf)
+{
+	static struct ldmsd_sampler plugin = {
+		.base = {
+			 .name = SAMP,
+			 .type = LDMSD_PLUGIN_SAMPLER,
+			 .term = term,
+			 .config = config,
+			 .usage = usage,
+			 },
+		.get_set = get_set,
+		.sample = sample,
+	};
+
+	log_fn = pf;
+	log_fn(LDMSD_LDEBUG,
+	       SAMP " get_plugin() called (" PACKAGE_STRING ")\n");
+
+	return &plugin.base;
+}
+
+/*********** let's count the vdev here *************/
+static int get_leaf_vdevs_count(nvlist_t * nvroot, const char *pool_name)
+{
+	uint_t children;
+	nvlist_t **child;
+
+	if (nvlist_lookup_nvlist_array(nvroot, ZPOOL_CONFIG_CHILDREN,
+				       &child, &children) == 0) {
+		for (int c = 0; c < children; c++) {
+			get_leaf_vdevs_count(child[c], pool_name);
+		}
+	} else {
+		/**vdevscount++;*/
```

Comment:
I don't think is comment adds anything helpful, I'd drop it.

Reply:
Dropped